### PR TITLE
Liquidator bot: Optimize calls and bot execution

### PR DESCRIPTION
**Pull request by chrismaree**
This PR optimizes the liquidator bot async calls and execution logic. 

## Scope of changes:
1) **ExpiringMultiPartyClient.js optimizations**: This follows on from @nicholaspai's PR #1805 by reducing the number of state calls and running more queries in parallel. In particular, the client now queries the `NewSponsor`, `EndedSponsorPosition`, and `LiquidationCreated` to greatly reduce the number of individual calls that are needed. By removing addresses from `EndedSponsorPosition` from `NewSponsor`, the client is able to only query current _active_ sponsors. This removes querying all ended sponsor positions.
2) **`liquidator.js` optimizations**: Previously, the liquidator had two main functions: `queryAndLiquidate` and `queryAndWithdrawRewards`. Each of these functions starts off by calling the local `_update` function that fetches the latest state from the EMP client and then running state queries. This PR refactors this behavior by pulling the update logic into the bot's `index.js` to ensure that the state is only updated _once_ before running any liquidation or rewards logic. This change immediately saves the number of queries by half.
3) **`index.js` optimizations**: the main entry point in the bot has a number of calls that can be performed in parallel. These changes decreased the total bot boot and load time.

## Quantifying performance improvements:
A reverse Proxy was set up to quantify the performance improvements due to this PR's refactors. The proxy was set up to sit between Infura and the local bot execution such that all traffic was re-routed via the proxy such that networking would flow as :`infura -> reverse proxy -> liquidator bot`. Through this configuration, all networking packages can be scraped and total data transferred can be quantified. This technique has shown to report the _exact_ number of queries that Infura would receive on a given polling cycle.

To perform a realistic test scenario the liquidator bot was run on **main net against the current production contract yUSD** which currently have ~200 active sponsors and historically had around ~400 unique users. The testing was performed in both serverless mode and loop mode.

### Severless mode
The changes in this PR resulted in a ~4x speedup in execution time and decreased the number of Infra calls from 2913 to 1201 (~2.4x less). Severless network traffic comparison can be seen in the screenshots below:
image

### Loop Mode
The changes from this PR are _really_ felt in the loop mode of execution. This is particularly relevant as most users will be running their bots in this mode if they were to run them on their own machine, GCP or AWS. 

It is important to note is that the initial bot booting from a cold start _places the same Infura load as one serverless loop_ (I verified this through the reverse proxy). This is relevant as we can use the serverless metrics to calculate what the _running load_ is of the bots. I.e how much Infura load do the bots produce when running in their "steady state" and polling every `POLLING_DELAY` interval.

Each new loop after this PR's refinements results in an additional 213 queries to Infura to update the local state. Before this PR, the bots previously would place 1631 _every loop_ to update their local state. This results in a reduction of ~7.6 times on normal operation!

The screenshot below compares the total load after one boot loop, then one full looping cycle.
image

## Extrapolating usage
Seeing the liquidator bots now require ~250 (adding a buffer from the actual number) calls every minute (assuming 60-second polling delay) they will consume 360 000 Infura calls per day. This is a measurable improvement from the current bots that consume 2 348 640 per day for _just_ the liquidator bot (note our actual usage is much lower than this because we don't poll at a 1-minute interval). While 360 000 is still above the free Infura tear, we could comfortably set the bot default polling delay to once every 5 minutes which would place the bot at ~72 000 calls per day which is **within the free quota**.

## Future work and testing
Next up is to improve the disputer and monitor bots with the same kind of reactors. I also want to run a liquidator bot against a fresh Infura key to validate the exact assumptions and calculates from the Extrapolating usage section.

### CODE REVIEW

File: financial-templates-lib/clients/ExpiringMultiPartyClient.js [hunk 1]
```diff
@@ -83,37 +83,59 @@ class ExpiringMultiPartyClient {
...
     this.cumulativeFeeMultiplier = this.toBN(cumulativeFeeMultiplier.toString());
+  }
+
+  async update() {
+    // If it is the first run then get contract contestants. This only needs to be called once.
+    if (!this.collateralRequirement) {
+      await this.initialSetup();
+    }
+    // Fetch contract state variables in parallel.
+    const [newSponsorEvents, endedSponsorEvents, liquidationCreatedEvents, currentTime] = await Promise.all([
```
Comment (nicholaspai) [inline financial-templates-lib/clients/ExpiringMultiPartyClient.js:111]: Drive-by, why are you querying these events now? Shouldn't these be handled by the EMPEventClient (which also needs refactoring)?

File: financial-templates-lib/clients/ExpiringMultiPartyClient.js [hunk 2]
```diff
@@ -83,37 +83,69 @@ class ExpiringMultiPartyClient {
...
+      const index = sponsorsToRemove.indexOf(address);
+      if (index !== -1) sponsorsToRemove = [...sponsorsToRemove.slice(0, index), ...sponsorsToRemove.slice(index + 1)];
+      return shouldRemoveSponsor;
+    });
+
+    // Array of all liquidated sponsors, over all time. Use a Set to ensure only contains unique elements.
+    const liquidatedSponsors = [...new Set(liquidationCreatedEvents.map(e => e.returnValues.sponsor))];
+
+    // Fetch sponsor position & liquidation in parallel batches, 20 at a time, to be safe and not overload the web3 node.
+    const WEB3_CALLS_BATCH_SIZE = 150;
```
Comment (chrismaree) [inline financial-templates-lib/clients/ExpiringMultiPartyClient.js:141]: I increased this to 150. I'm pretty sure the infura nodes can take this [no problem](https://ethereum.stackexchange.com/questions/23177/how-many-concurrent-requests-can-an-ethereum-node-handle). If we bump this up we'll reduce the run time even further.

File: financial-templates-lib/clients/ExpiringMultiPartyClient.js [hunk 3]
```diff
@@ -83,37 +83,69 @@ class ExpiringMultiPartyClient {
...
+      this.emp.getPastEvents("LiquidationCreated", { fromBlock: 0 }),
+      this.emp.methods.getCurrentTime().call()
+    ]);
+
+    // Array of all sponsors, over all time. Can contain repeats for every `NewSponsor` event.
+    const newSponsorAddresses = [...newSponsorEvents.map(e => e.returnValues.sponsor)];
+
+    // Array of ended sponsors. Can contain repeats for every `EndedSponsorPosition` event.
+    const endedSponsorAddresses = [...endedSponsorEvents.map(e => e.returnValues.sponsor)];
+
```
Comment (chrismaree) [inline financial-templates-lib/clients/ExpiringMultiPartyClient.js:123]: The logic below is a bit tricky and took me a while to get just right. The basic idea is outlined in the comments. Overall the idea is to only run the `this.emp.methods.positions` call against **active** sponsors. I use a combination of `NewSponsor` and `EndedSponsorPosition` events to build an array of _active_ sponsors.

File: financial-templates-lib/clients/ExpiringMultiPartyClient.js (hunk 1 above)
Comment (chrismaree) [reply financial-templates-lib/clients/ExpiringMultiPartyClient.js:111]: take a look at the implementation below. I use the two sets of events to build an array of `ActiveSponsors`. I use the `EndedSponsorPosition` to filter out sponsors that have ended their position from the `NewSponsor` events. This means we don't have to query `this.emp.methods.positions` for sponsors that have ended their positions.

File: financial-templates-lib/clients/ExpiringMultiPartyClient.js [hunk 4]
```diff
@@ -157,26 +189,19 @@ class ExpiringMultiPartyClient {
     this.expiredLiquidations = expiredLiquidations;
     this.disputedLiquidations = disputedLiquidations;
```
Comment (chrismaree) [inline financial-templates-lib/clients/ExpiringMultiPartyClient.js:191]: I changed the below from the reduce to a map as there is no longer zero collateral positions due to the filtering above and the map is way easier to read.

File: financial-templates-lib/test/clients/ExpiringMultiPartyClient.js [hunk 5]
```diff
@@ -167,7 +166,7 @@ contract("ExpiringMultiPartyClient.js", function(accounts) {
 
     await updateAndVerify(
       client,
-      [sponsor1, sponsor2],
+      [sponsor1],
```
Comment (chrismaree) [inline financial-templates-lib/test/clients/ExpiringMultiPartyClient.js:169]: this test was actually **wrong** I believe. The sponsor is liquidated and so the address should _not_ be in the client any more when calling `client.getAllSponsors()` in `updateAndVerify`

File: financial-templates-lib/test/clients/ExpiringMultiPartyClient.js [hunk 6]
```diff
@@ -216,7 +215,68 @@ contract("ExpiringMultiPartyClient.js", function(accounts) {
...
+          sponsor: sponsor1,
+          numTokens: toWei("100"),
+          amountCollateral: toWei("20"),
+          hasPendingWithdrawal: false,
+          withdrawalRequestPassTimestamp: "0",
+          withdrawalRequestAmount: "0"
+        }
+      ]
+    );
+    // If sponsor, creates, redeemes and then creates again they should now appear in the table.
```
Comment (chrismaree) [inline financial-templates-lib/test/clients/ExpiringMultiPartyClient.js:251]: I added the tests below to double-check the removing logic in the complex filter within the `update` function.

File: liquidator/liquidator.js [hunk 7]
```diff
@@ -315,20 +318,15 @@ class Liquidator {
...
       });
     }
-
-    // Update the EMP Client since we created new liquidations.
-    await this.empClient.update();
   }
 
   // Queries ongoing liquidations and attempts to withdraw rewards from both expired and disputed liquidations.
-  async queryAndWithdrawRewards() {
+  async withdrawRewards() {
```
Comment (chrismaree) [inline liquidator/liquidator.js:324]: I renamed this function as it no longer does any `update`. that's left for the `index.js`

File: liquidator/test/Liquidator.js [hunk 8]
```diff
@@ -163,7 +163,8 @@ contract("Liquidator.js", function(accounts) {
     // Start with a mocked price of 1 usd per token.
     // This puts both sponsors over collateralized so no liquidations should occur.
     priceFeedMock.setCurrentPrice(toBN(toWei("1")));
-    await liquidator.queryAndLiquidate();
+    await liquidator.update();
```
Comment (chrismaree) [inline liquidator/test/Liquidator.js:166]: tests only changed to accommodate the new function signature and to explicitly call `update` before hand.

File: financial-templates-lib/clients/ExpiringMultiPartyClient.js [hunk 9]
```diff
@@ -83,37 +83,69 @@ class ExpiringMultiPartyClient {
...
+      this.emp.getPastEvents("EndedSponsorPosition", { fromBlock: 0 }),
+      this.emp.getPastEvents("LiquidationCreated", { fromBlock: 0 }),
+      this.emp.methods.getCurrentTime().call()
+    ]);
+
+    // Array of all sponsors, over all time. Can contain repeats for every `NewSponsor` event.
+    const newSponsorAddresses = [...newSponsorEvents.map(e => e.returnValues.sponsor)];
+
+    // Array of ended sponsors. Can contain repeats for every `EndedSponsorPosition` event.
+    const endedSponsorAddresses = [...endedSponsorEvents.map(e => e.returnValues.sponsor)];
```
Comment (mrice32) [inline financial-templates-lib/clients/ExpiringMultiPartyClient.js:122]: tiny nit; can these be slightly simplified to:
```
const newSponsorAddresses = newSponsorEvents.map(e => e.returnValues.sponsor);
```
and 
```
const endedSponsorAddresses = endedSponsorEvents.map(e => e.returnValues.sponsor);
```
?

File: financial-templates-lib/clients/ExpiringMultiPartyClient.js [hunk 10]
```diff
@@ -83,37 +83,69 @@ class ExpiringMultiPartyClient {
...
+
+    // Filter out the active sponsors by removing ended sponsors from all historic sponsors. Note that a sponsor could
+    // create a position, end their position and then create another one. They could potentially do this multiple times.
+    // As a result, create a temp array `sponsorsToRemove` which is updated as sponsors are removed from the array.
+    let sponsorsToRemove = endedSponsorAddresses; // temp array to contain address to remove.
+    this.activeSponsors = newSponsorAddresses.filter(address => {
+      // If the sponsorsToRemove contains the current address, then that address should be removed from newSponsorAddresses.
+      const shouldRemoveSponsor = !sponsorsToRemove.includes(address);
+      // Update the `sponsorsToRemove` by removing the first instance of the address from the sponsorsToRemove array.
+      const index = sponsorsToRemove.indexOf(address);
```
Comment (mrice32) [inline financial-templates-lib/clients/ExpiringMultiPartyClient.js:132]: nit: do you need the `shouldRemoveSponsor` line here? Shouldn't the `indexOf` being `-1` or not tell you whether you need to remove it or not?

File: financial-templates-lib/clients/ExpiringMultiPartyClient.js [hunk 11]
```diff
@@ -83,37 +83,69 @@ class ExpiringMultiPartyClient {
...
+      if (index !== -1) sponsorsToRemove = [...sponsorsToRemove.slice(0, index), ...sponsorsToRemove.slice(index + 1)];
+      return shouldRemoveSponsor;
```
Comment (mrice32) [inline financial-templates-lib/clients/ExpiringMultiPartyClient.js:134]: nit: `shouldRemoveSponsor` seems like the opposite of the name you want -- shouldn't it be `shouldKeepSponsor`?

File: financial-templates-lib/clients/ExpiringMultiPartyClient.js [hunk 12]
```diff
@@ -83,37 +83,69 @@ class ExpiringMultiPartyClient {
...
     this.cumulativeFeeMultiplier = this.toBN(cumulativeFeeMultiplier.toString());
+  }
 
-    // Fetch sponsor position, liquidation, and current time data in parallel batches, 20 at a time, to be safe and not overload the web3 node.
-    const WEB3_CALLS_BATCH_SIZE = 20;
-    const [positions, allLiquidations, currentTime] = await Promise.all([
-      Promise.map(this.sponsorAddresses, address => this.emp.methods.positions(address).call(), {
+  async update() {
+    // If it is the first run then get contract contestants. This only needs to be called once.
+    if (!this.collateralRequirement) {
```
Comment (nicholaspai) [inline financial-templates-lib/clients/ExpiringMultiPartyClient.js:107]:
```suggestion
    if (!this.collateralRequirement || !this.liquidationLiveness || !this.cumulativeFeeMultiplier) {
```

File: financial-templates-lib/clients/ExpiringMultiPartyClient.js [hunk 13]
```diff
@@ -157,26 +189,19 @@ class ExpiringMultiPartyClient {
...
-              numTokens: positions[i].tokensOutstanding.toString(),
-              amountCollateral: rawCollateral
-                .mul(this.cumulativeFeeMultiplier)
-                .div(this.toBN(this.toWei("1")))
-                .toString(),
-              hasPendingWithdrawal: positions[i].withdrawalRequestPassTimestamp > 0
-            }
-          ]);
-    }, []);
+    this.positions = activePositions.map((position, index) => {
```
Comment (nicholaspai) [inline financial-templates-lib/clients/ExpiringMultiPartyClient.js:192]: Much cleaner final mapping +1

File: financial-templates-lib/test/clients/ExpiringMultiPartyClient.js (hunk 6 above)
Comment (nicholaspai) [reply financial-templates-lib/test/clients/ExpiringMultiPartyClient.js:251]: Perfect, I was going to ask about testing the NewSponsor <--> EndedSponsor timeline

File: liquidator/index.js [hunk 14]
```diff
@@ -90,13 +90,19 @@ async function run(logger, address, pollingDelay, priceFeedConfig, liquidatorCon
 
     // The EMP requires approval to transfer the liquidator's collateral and synthetic tokens in order to liquidate
     // a position. We'll set this once to the max value and top up whenever the bot's allowance drops below MAX_INT / 2.
-    await gasEstimator.update();
-    const collateralToken = await ExpandedERC20.at(await emp.collateralCurrency());
-    const syntheticToken = await ExpandedERC20.at(await emp.tokenCurrency());
-    const currentCollateralAllowance = await collateralToken.allowance(accounts[0], empClient.empAddress);
-    const currentSyntheticAllowance = await syntheticToken.allowance(accounts[0], empClient.empAddress);
+    const [collateralToken, syntheticToken] = await Promise.all([
```
Comment (nicholaspai) [inline liquidator/index.js:93]: Should we batch these with the above Promise.all?

File: financial-templates-lib/clients/ExpiringMultiPartyClient.js (hunk 10 above)
Comment (nicholaspai) [reply financial-templates-lib/clients/ExpiringMultiPartyClient.js:132]: This is a good point, I think this would improve readability as this `.filter()` is the most complex change in this PR

Comment (chrismaree) [reply financial-templates-lib/clients/ExpiringMultiPartyClient.js:132]: ye, this is a good point. I can use `indexOf` to infer if it `includes` and joint them together.

File: financial-templates-lib/clients/ExpiringMultiPartyClient.js (hunk 11 above)
Comment (chrismaree) [reply financial-templates-lib/clients/ExpiringMultiPartyClient.js:134]: I scrapped this variable in its entirety and simply used the implications from the `indexOf`. PTAL and let me know if it's easier to read like this.

File: liquidator/index.js (hunk 14 above)
Comment (chrismaree) [reply liquidator/index.js]: ye, can do & done.

**Comment by chrismaree** [comment]
@mrice32 & @nicholaspai I've implemented your nits. PTAL

File: financial-templates-lib/clients/ExpiringMultiPartyClient.js (hunk 2 above)
Comment (nicholaspai) [reply financial-templates-lib/clients/ExpiringMultiPartyClient.js:143]: Cool with me I have no opinion here. Perhaps @daywiss has a better idea
